[DISCO-2178] Version Contract Tests for all environments

taddes opened the pull request:
# Description
* Version endpoint contract tests. 
* Took alternative path to cover all environments and to not simply skip in local. Achieved through:
- regex validation of sha-1 hash
- type check for HttpUrl to verify type check variance
- empty version string once in stage/prod

# Issue(s)
#170 | [DISCO-2178](https://mozilla-hub.atlassian.net/browse/DISCO-2178)

Replaces [old branch](https://github.com/mozilla-services/merino-py/pull/205) that was created when dealing with blocking cache bug that was resolved [DISCO-2264](https://mozilla-hub.atlassian.net/browse/DISCO-2264)



[DISCO-2178]: https://mozilla-hub.atlassian.net/browse/DISCO-2178?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DISCO-2264]: https://mozilla-hub.atlassian.net/browse/DISCO-2264?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Trinaa commented on tests/contract/client/tests/test_merino.py:
```diff
@@ -114,12 +118,22 @@ def merino_step(step: Step) -> None:
+                    step_content=step.response.content,  # type: ignore
+                    merino_version_content=VersionResponseContent(**response.json()),
+                )
+                return
+            else:
```
Returning in the if statement makes the else statement redundant. You could either eliminate the else, or extract the return from both if and else clauses.

taddes replied:
The intended behavior is to make the else statement redundant, since there are 2 types of Response content, one for all the other Merino responses and this one for the version endpoint.  If a 200 returns, the first check triggers whether or not it's the case of being for the `version` endpoint. After the version assertion runs, we don't want there to be another check with the wrong model. The original code for the base 200 assertion function had a return right after it like before.

`python

        if response.status_code == 200:
            # If the response status code is 200 OK, use the
            # assert_200_response() helper function to validate the content of
            # the response from Merino. This includes creating a pydantic model
            # instance for checking the field types and comparing a dict
            # representation of the model instance with the expected response
            # content for this step in the test scenario.
            assert_200_response(
                step_content=step.response.content,
                merino_content=ResponseContent(**response.json()),
                fetch_kinto_icon_url=fetch_kinto_icon_url,
            )
            return`

Trinaa commented on tests/contract/client/tests/test_merino.py:
```diff
@@ -114,12 +118,22 @@ def merino_step(step: Step) -> None:
-            )
-            return
+
+            if step.request.path == "/__version__":
```
I suggest to change this condition to use the expected response type instead of the request path

```suggestion
            if type(step.response.content) == VersionResponseContent:
```

Trinaa reviewed (changes requested):
Hi @taddes, thank you for contributing this test!!!

I will leave the final implementation up to you, but my 2 cents is I would have preferred a 'skip' feature to this implementation, which is very specific to the needs of the version endpoint tests. I believe the 'skip' would have future benefit and don't think there is added benefit to running the __version__ test locally.

taddes commented:
Hey @Trinaa ! Thanks for the review comments. As requested, I've done the following

- Added docker command to Makefile, great idea! 
- Modified docs for clarity and put Make-related stuff in the upper level as you thought.
- Removed that unnecessary `return` stmt
- Re: skip logic, as we discussed in Slack, I'm gonna leave things be so we can at least verify local results since I wasn't a fan of putting misleading data in the scenarios file. Also, I like the gut check provided by us verifying the shape of data in stage/prod all while also having a local verification. We can definitely implement skip logic in the future should we wish, since this revealed for us how to scope and extract environment variables that were previously not accessible to us. Therefore gonna leave the scenario data as-is, since it conforms to this pattern. 
- Re: `  if type(step.response.content) == VersionResponseContent:` for now, I added this as an and to the initial `__version__` path because just doing that single check makes Pydantic mad when it comes to the `wiki_fruit` tests and can't grok them, so wasting time fighting with those just isn't worth the time. I like the premise of checking the type. This way, we do that and also verify the path, further scoping the version test. Hope that's fine with you. 
```python
            if (
                step.request.path == "/__version__"
                and type(step.response.content) == VersionResponseContent
            ):```

Otherwise, all resolved and good to go if you are!

Trinaa reviewed (approved):
Okay. Lets merge. 🚀
